examples/codes/external_code.ex.py

```diff
@@ -0,0 +1,387 @@
+
+    def _read_file(self):
+        out_params = {}
+        with open(self.outfile, "r") as of:
+            for line in of:
+                if line.startswith("#"):
+                    pass
+                if line.startswith(" "):
+                    k, v = line.split()
+                    out_params[k] = float(v)
+        self._update_params_with_outputs(out_params)
+
+    def run(self) -> ParameterFrame:
+        """Run mode"""
+        self._read_file()
+        return self.params
+
+    def read(self) -> ParameterFrame:
+        """Read mode"""
+        self._read_file()
+        return self.params
+
+    def mock(self) -> ParameterFrame:
+        """Mock mode"""
+        self._update_params_with_outputs({"param1": 15})
+        return self.params
+
+
+# %% [markdown]
+# `Solver` combines the three tasks into one object for execution.
+# The execute method has been overridden here for our usecase and returns
+# the `ParameterFrame`.
+
+# %%
+class Solver(CodesSolver):
+    """The External Code Solver."""
+
+    name = PRG_NAME
+    params = ECParameterFrame
+    setup_cls = Setup
+    run_cls = Run
+    teardown_cls = Teardown
+    run_mode_cls = RunMode
+
+    def __init__(self, params: Union[ParameterFrame, Dict], build_config: Dict):
+
+        self.params = ECParameterFrame.from_defaults()
+
+        if isinstance(params, ParameterFrame):
+            self.params.update_from_frame(params)
+        else:
+            try:
+                self.params.update_from_dict(params)
+            except TypeError:
+                self.params.update_values(params)
+        self._setup = self.setup_cls(
+            self.params, build_config.get("problem_settings", {}), build_config["infile"]
+        )
+        self._run = self.run_cls(
+            self.params,
+            build_config["infile"],
+            build_config["outfile"],
+            build_config.get("binary", BINARY),
+        )
+        self._teardown = self.teardown_cls(self.params, build_config["outfile"])
+
+    def execute(self, run_mode: Union[str, RunMode]) -> ParameterFrame:
+        """Execute the solver"""
+        if isinstance(run_mode, str):
+            run_mode = self.run_mode_cls.from_string(run_mode)
+        result = None
+        if setup := self._get_execution_method(self._setup, run_mode):
+            result = setup()
+        if run := self._get_execution_method(self._run, run_mode):
+            run(result)
+        if teardown := self._get_execution_method(self._teardown, run_mode):
+            result = teardown()
+        return result
+
+
+# %% [markdown]
+# ### Using the solver
+#
+# To run the solver you just need to provide the parameters and the configuration
+# to initialse the object.
+# Be aware `problem_settings` should be used sparingly for options that won't
+# change within the rerunning of the solver,
+# it has the same effect as modifying the default.
+# Also note the units of the parameter values returned back have been updated
```

Comment:
We get a few warnings when we run this example. Is it worth mentioning them here, or in a comment within the code block? Just to let the users know this is expected and why.

```suggestion
# Also note the units of the parameter values returned back have been updated.
```